### DISCUSSION TITLE
Content -> lcdmenu.forntoh.dev

### DISCUSSION THREAD

**Discussion by maansch** [forntoh/LcdMenu#117] [Q&A] [answered]
Hi,
before you changed the documentation for version V3.3.0 (ItemProgress), there were different overviews with descriptions.
e.g.
LcdMenu

> The LcdMenu class contains all fields and methods to manipulate the menu items.
> Private Fields
> 	uint8_tcursorPosition =1
> 	Cursor position
> 	uint8_ttop =1
> 	First visible item’s position in the menu array
> 	boolisEditModeEnabled =false
> 	Edit mode
> 	boolisEditModeEnabled =false
> 	Will prevent left and right movement when in edit mode and character picker is active
> 	uint8_t bottom = 0
> 	Last visible item’s position in the menu array
> 	uint8_tmaxRows
> 	Rows on the LCD Display
> 	uint8_tmaxCols
> 	Columns on the LCD Display
> 	uint8_tblinkerPosition =0
> 	Colum location of Blinker
> 	MenuItem* currentMenuTable =NULL
> 	An array of menu items
> 	byte downArrow[8]
> 	Down arrow (↓)
> 	byte upArrow[8]
> 	Up arrow (↑)
> 	uint8_tcursorIcon =0x7E
> 	Cursor icon. Defaults to the right arrow (→).
> 	boolenableUpdate =true
> 
>     Determines whether the screen should be updated after an action. Set it to false when you want to display any other content on the screen then set it back to true to show the menu.
> 
> Private Methods
> 	voiddrawCursor()
> 	Draws the cursor
> 	voiddrawMenu()
> 	Draw the menu items with up and down indicators
> 	booleanisAtTheStart()
> 	Check if the cursor is at the start of the menu items
> 	Returns:
> 
>     true: boolean if it is at the start
> 
> 	booleanisAtTheEnd()
> 	Check if the cursor is at the end of the menu items
> 	Returns:
> 
>     true: boolean if it is at the end
> 
> 	voidreset(booleanisHistoryAvailable)
> 	Reset the display
> 	Params:
> 
>     isHistoryAvailable - indicates if there is a previous position
> 
> voidresetBlinker()
> 
>     Calculate and set the new blinker position
> 
> Public Fields
> unsigned long startTime =0
> The time when the timer started in milliseconds
> unsigned int duration =0
> How long should the timer delay in milliseconds
> LiquidCrystal_I2C* lcd or LiquidCrystal* lcd
> 
>     LCD Display
> 
> Constructor
> Constructor for the LcdMenu class
> LcdMenu(uint8_t maxRows, uint8_t maxCols)
> Params:
> maxRows - rows on LCD display e.g. 4
> 
>     maxCols - columns on LCD display e.g. 20
> 
> Returns:
> 
>     new LcdMenu object
> 
> Public Methods
> Call this function in setup() to initialize the LCD and the custom characters used as up and down arrows
> Params:
> lcd_Addr - address of the LCD on the I2C bus (default 0x27)
> 
>     menu - menu to display
> 
> void update()
> 	Draw the menu items and cursor
> 	
> 	Call this function to set sub-menu items for any main menu item
> void setSubMenu(uint8_t position, MenuItem* items)
> 	Params:
> 	position - main menu item/where to place the sub-menu
> 
>     items - sub menu items

MenuItem

> The MenuItem class
> Public Fields
>   boolean isOn = false
> Boolean state of the item (either ON or OFF)
>   String value
> String value of an ItemInput
>   uint8_t itemIndex = 0
> Current index of list for ItemList
>   uint8_t itemCount = 0
> 
>     Number of items in the list for ItemList
> 
> Getters
> getText
> 	Get the text of the item
> 	const char* getText() { return text; }
> 	Returns:
> 
>     String - Item’s text
> 
> getCallback
> 	Get the callback of the item
> 	fptr getCallback() { return callback; }
> 	Returns:
> 
>     ftpr - Item’s callback
> 
> getCallbackInt
> 	Get the callback of the item
> 	fptrInt getCallbackInt() { return callbackInt; }
> 	Returns:
> 
>     fptrInt - Item’s callback
> 
> getCallbackStr
> 	Get the callback of the item
> 	fptrStr getCallbackStr() { return callbackStr; }
> 	Returns:
> 
>     fptrStr - Item’s callback
> 
> getSubMenu
> 	Get the sub-menu at item
> 	MenuItem* getSubMenu() { return subMenu; }
> 	Returns:
> 
>     MenuItem* - Submenu at item
> 
> 	getType
> 	Get the type of the item
> 	byte getType() { return type; }
> 	Returns:
> 
>     byte - type of menu item
> 
> getTextOn
> Get the text when toggle is ON
>  char* getTextOn() { return textOn; }
> Returns:
> 
>     String - ON text
> 
> getTextOff
> Get the text when toggle is OFF
>  char* getTextOff() { return textOff; }
> Returns:
> 
>     String - OFF text
> 
> getItems
> 	Get the list of items
> 	String* getItems() { return items; }
> 	Returns:
> 
>     String* - List of items
> 
> Setters
> setText
> Set the text of the item
>  void setText(const char* text) { this->text = text; }
> Params:
> 
>     text - text to display for the item
> 
> setCallBack
> Set the callback on the item
>  void setCallBack(fptr callback) { this->callback = callback; }
> Params:
> 
>     callback - reference to the callback function
> 
> setSubMenu
> 	Set the sub-menu on the item
> 	void setSubMenu(MenuItem* subMenu) { this->subMenu = subMenu; }
> 	Params:
> 
>     subMenu - for the item
> 
> Operators
> Get item at index from the submenu
>  MenuItem& operator[](const uint8_t index)
> Params:
> 
>     index - for the item
> 
> Extension Classes
> ItemHeader
> This item must be present in all menus collections (the first item in the array).
> Example
> MenuItem mainMenu[] = {ItemHeader(),
>                        MenuItem("Item 1"),
>                        MenuItem("Item 2"),
>                        ...
> Constructors
> ItemHeader(MenuItem* parent)
> Params:
> 
>     parent - the parent menu item
> 
> ItemFooter
> This item must be present in all menus (the last item in the array)
> Example
> MenuItem mainMenu[] = {ItemHeader(),
>                        MenuItem("Item 1"),
>                        MenuItem("Item 2"),
>                        ...
>                        ItemFooter()};
> Constructor
> ItemFooter()
> ItemInput
> This is an item type where a user can type in information, the information is persisted in the item and can be gotten later by using item->value
> Constructors
> ItemInput(char* text, fptrStr callback)
> Params:
> text - text to display for the item
> value - the input value
> 
>     callback - reference to the callback function
> 
> ItemSubMenu
> 	This item type indicates that the current item contains a sub-menu. The sub-menu is opened whenenter() is invoked.
> 	Params:
> 	text - text to display for the item
> 
>     parent - the parent of the sub-menu item
> 
> 	ItemSubMenu(const char* text, MenuItem* parent)
> 	
> 	ItemToggle
> 	This item type indicates that the current item is toggleable. When enter() is invoked, the state of isOn is toggled.
> 	Constructor
> 	ItemToggle(const char* key, char* textOn, char* textOff, fptrInt callback)
> 	Params:
> 	key - key of the item
> 	textOn - display text when ON
> 	textOff - display text when OFF
> 
>     callback - reference to the callback function
> 
> ItemCommand
> This item type indicates that the current item is a command. When enter() is invoked, the command (callback) bound to this item is invoked.
> Constructor
> ItemCommand(const char* key, fptr callback)
> Params:
> key - key of the item
> 
>     callback - reference to the callback function
> 
> ItemList
> This item type indicates that the current item is a list.
> When left() is invoked the view cycles down the list
> When right() is invoked the view cycles up the list, you can use only right() if you have a single button, because once the menu reaches the end of the list, it automatically starts back from the begining.
> 
>     When enter() is invoked, the command (callback) bound to this item is invoked.
> 
> This can be used for other primitive data types, you just need to pass it as string then parse the result to the desired datatype
> Constructor
> ItemList(const char* key, String* items, uint8_t itemCount, fptrInt callback)
> Params:
> key - key of the items
> items - items to display
> itemCount - number of items in items
> 
>     callback - reference to the callback function
> 

etc

I found it very helpful, can you please unlock it again?
br
Marc-Andre

PS:
In various of your answers you refer to pages that no longer exist.
For example here: calling getSubmenu()
https://github.com/forntoh/LcdMenu/discussions/14#discussioncomment-1341805

**Comment by forntoh** [comment, marked as answer]
Since the code is properly documented, and the overview was just a copy of the code while stripping the content, I don't plan to do this as it takes effort to maintain whenever there are changes, if you want to you can do it, here's the branch for the documentation https://github.com/forntoh/LcdMenu/tree/docs/docs

**Reply by forntoh** [reply]
I have also updated the reference in that comment
